Sources/Citadel/OpenSSHKey.swift

```diff
@@ -0,0 +1,223 @@
+    }
+}
+
+extension ED25519.PrivateKey: ReadableFromBuffer {
+    static func read(from buffer: inout ByteBuffer) throws -> Self {
+        guard
+            let publicKeyLength = buffer.readInteger(as: UInt32.self),
+            let publicKey = buffer.readBytes(length: Int(publicKeyLength))
+        else {
+            throw InvalidKey()
+        }
+
+        guard
+            let privateKeyLength = buffer.readInteger(as: UInt32.self),
+            let privateKey = buffer.readBytes(length: Int(privateKeyLength)),
+            Array(privateKey[32...]) == publicKey
+        else {
+            throw InvalidKey()
+        }
+        
+        return try Self.init(rawRepresentation: privateKey[..<32])
+    }
+}
+
+enum OpenSSH {
+    enum Cipher: String {
+        case none
+    }
+    
+    enum KDF: String {
+        case none
+    }
+    
+    enum KeyType: String {
+        case sshRSA = "ssh-rsa"
+        case sshED25519 = "ssh-ed25519"
+    }
+    
+    struct PrivateKey<SSHKey: OpenSSHKeyProtocol> {
+        let cipher: Cipher
+        let kdf: KDF
+        let numberOfKeys: Int
+        let publicKey: SSHKey.PublicKey
+        let privateKey: SSHKey.PrivateKey
+        let comment: String
+        
+        var keyType: KeyType {
+            SSHKey.keyType
+        }
+    }
+}
+
+extension OpenSSH.PrivateKey {
+    init(string key: String) throws {
+        var key = key.replacingOccurrences(of: "\n", with: "")
+        
+        guard
+            key.hasPrefix("-----BEGIN OPENSSH PRIVATE KEY-----"),
+            key.hasSuffix("-----END OPENSSH PRIVATE KEY-----")
+        else {
+            throw InvalidKey()
+        }
+        
+        key.removeLast("-----END OPENSSH PRIVATE KEY-----".utf8.count)
+        key.removeFirst("-----BEGIN OPENSSH PRIVATE KEY-----".utf8.count)
+        
+        guard let data = Data(base64Encoded: key) else {
+            throw InvalidKey()
+        }
+        
+        var buffer = ByteBuffer(data: data)
+        
+        guard
+            buffer.readString(length: "openssh-key-v1".utf8.count) == "openssh-key-v1",
+            buffer.readInteger(as: UInt8.self) == 0x00
+        else {
+            throw InvalidKey()
+        }
+            
+        self.cipher = try OpenSSH.Cipher(from: &buffer)
+        self.kdf = try OpenSSH.KDF(from: &buffer)
+        
+        guard
+            let numberOfKeys = buffer.readInteger(as: UInt32.self),
+            numberOfKeys == 1 // # of keys always one
+        else {
+            throw InvalidKey()
+        }
+        
+        self.numberOfKeys = Int(numberOfKeys)
+        
+        guard
+            let publicKeyBufferLength = buffer.readInteger(as: UInt32.self).map(Int.init),
+            var publicKeyBuffer = buffer.readSlice(length: publicKeyBufferLength)
+        else {
+            throw InvalidKey()
+        }
+        
+        let publicKeyType = try OpenSSH.KeyType(from: &publicKeyBuffer)
```

Comment:
Assuming you implement the above `decryption` remark, we'll also need to decrypt the `publicKeyBuffer` and `privateKeyBuffer` using that decrypt method.